gc/base/standard/ConcurrentGC.cpp

```diff
@@ -174,19 +174,21 @@ con_helper_thread_proc2(OMRPortLibrary* portLib, void *info)
 } /* extern "C" */
 
 void
-MM_ConcurrentGC::reportGCCycleStart(MM_EnvironmentStandard *env)
+MM_ConcurrentGC::reportGCCycleStart(MM_EnvironmentBase *env)
 {
 	/* TODO CRGTMP - Temporarily move this code to the point when
 	 * we have exclusiveVMAccess so that GC check works.  Move it back
 	 * once we complete JAZZ 33760.
 	 */
+	MM_CycleState *previousCycleState = env->_cycleState;
+
 	_concurrentCycleState = MM_CycleState();
 	env->_cycleState = &_concurrentCycleState;
 	env->_cycleState->_type = _cycleType;
 
 	MM_ParallelGlobalGC::reportGCCycleStart(env);
 
-	env->_cycleState = NULL;
+	env->_cycleState = previousCycleState;
```

Comment:
So after backing out this change to I see this (segfault) again. This particular code path predates my changes and would previously have resulted in segfault whenever MM_ConcurrentGC::internalPreCollect() was called prior to concurrent kickoff. 

Only explanation I can see is that this branch (at ConcurrentGC.cpp line 2944, see gdb trace below) is not normally taken and my changes have created this condition, at least for this test case. So my changes are delaying or preventing concurrent kickoff. Looking into that, but I think this code will need fixing in any case -- this branch will always segfault as it stands.

GC policy for trace below was optavgpause.

```
Breakpoint 17, MM_ConcurrentGC::internalPreCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, subSpace=0x7ffff0107640, allocDescription=0x7ffff79a0210, gcCode=0) at ConcurrentGC.cpp:2942
2942            if (executionModeAtGC == CONCURRENT_OFF) {
(gdb) bt
#0  MM_ConcurrentGC::internalPreCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, subSpace=0x7ffff0107640, allocDescription=0x7ffff79a0210, gcCode=0) at ConcurrentGC.cpp:2942
#1  0x00007ffff4499ba4 in MM_Collector::preCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, subSpace=0x7ffff0107640, allocDescription=0x7ffff79a0210, gcCode=0) at Collector.cpp:202
#2  0x00007ffff449a059 in MM_Collector::garbageCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, callingSubSpace=0x7ffff0107640, allocateDescription=0x7ffff79a0210, gcCode=0, objectAllocationInterface=0x0, baseSubSpace=0x7ffff0107290, context=0x0) at Collector.cpp:301
#3  0x00007ffff46066c6 in MM_MemorySubSpaceFlat::allocationRequestFailed (this=0x7ffff0107640, env=0x7ffff00426e8, allocateDescription=0x7ffff79a0210, allocationType=MM_MemorySubSpace::ALLOCATION_TYPE_OBJECT, objectAllocationInterface=0x0, baseSubSpace=0x7ffff0107290, previousSubSpace=0x7ffff0107290) at MemorySubSpaceFlat.cpp:143
#4  0x00007ffff4608f92 in MM_MemorySubSpaceGeneric::allocateObject (this=0x7ffff0107290, env=0x7ffff00426e8, allocDescription=0x7ffff79a0210, baseSubSpace=0x0, previousSubSpace=0x0, shouldCollectOnFailure=true) at MemorySubSpaceGeneric.cpp:303
#5  0x00007ffff4617a88 in MM_TLHAllocationInterface::allocateObject (this=0x7ffff0136370, env=0x7ffff00426e8, allocDescription=0x7ffff79a0210, memorySpace=0x7ffff01079b0, shouldCollectOnFailure=true) at TLHAllocationInterface.cpp:194
#6  0x00007ffff4617c6a in MM_TLHAllocationInterface::allocateArrayletSpine (this=0x7ffff0136370, env=0x7ffff00426e8, allocateDescription=0x7ffff79a0210, memorySpace=0x7ffff01079b0, shouldCollectOnFailure=true) at TLHAllocationInterface.cpp:226
#7  0x00007ffff45cf484 in MM_ArrayletAllocationModel::allocateAndInitializeLayout (this=0x7ffff01365a0, memorySpace=0x7ffff01079b0, env=0x7ffff00426e8, allocDesc=0x7ffff79a0210, numberIndexedFields=1310720, shouldCollectOnFailure=true) at ArrayletAllocationModel.cpp:122
#8  0x00007ffff45d0fa0 in MM_ArrayletAllocationModel::allocateIndexableObject (this=0x7ffff01365a0, env=0x7ffff00426e8, allocDescription=0x7ffff79a0210, numberIndexedFields=1310720, shouldCollectOnFailure=true) at ArrayletAllocationModel.cpp:315
#9  0x00007ffff42aad65 in J9AllocateIndexableObject (vmThread=0x64b300, clazz=0x689c00, size=1310720, allocateFlags=1) at mgcalloc.cpp:516
#10 0x00007ffff6768994 in VM_BytecodeInterpreter::anewarray (this=0x7ffff79a0500, _sp=@0x7ffff79a0468, _pc=@0x7ffff79a0470) at BytecodeInterpreter.hpp:7283
#11 0x00007ffff6780678 in VM_BytecodeInterpreter::run (this=0x7ffff79a0500, vmThread=0x64b300) at BytecodeInterpreter.hpp:9846
#12 0x00007ffff674c9c1 in bytecodeLoop (currentThread=0x64b300) at BytecodeInterpreter.cpp:90
#13 0x00007ffff67f795e in cInterpreter () at javamisc.s:222
#14 0x0000000000000000 in ?? ()
(gdb) p *_stats
$5 = {
  <MM_Base> = {<No data fields>},
  members of MM_ConcurrentGCStats:
  _executionMode = 1,
  _executionModeAtGC = 1,
  _initWorkRequired = 73728,
  _traceSizeTarget = 1244869,
  _kickoffThreshold = 171628,
  _cardCleaningThreshold = 67610,
  _remainingFree = 0,
  _allocationsTaxed = 0,
  _allocationsTaxedAt0 = 0,
  _allocationsTaxedAt25 = 0,
  _allocationsTaxedAt50 = 0,
  _allocationsTaxedAt75 = 0,
  _allocationsTaxedAt100 = 0,
  _traceSizeCount = 0,
  _cardCleanCount = 0,
  _conHelperTraceSizeCount = 0,
  _conHelperCardCleanCount = 0,
  _completeTracingCount = 0,
  _finalTraceCount = 0,
  _finalCardCleanCount = 0,
  _RSScanTraceCount = 0,
  _RSObjectsFound = 0,
  _threadsScannedCount = 0,
  _threadsToScanCount = 0,
  _concurrentWorkStackOverflowOcurred = false,
  _concurrentWorkStackOverflowCount = 0,
  _completedModes = 0,
  _kickoffReason = NO_KICKOFF_REASON,
  _cardCleaningReason = CARD_CLEANING_REASON_NONE
}
(gdb) p _extensions->scavengerEnabled
$7 = false
(gdb) p _extensions->concurrentMark
$8 = true
(gdb) p _extensions->concurrentKickoffEnabled
$9 = true
(gdb) list 2942,2945
2942            if (executionModeAtGC == CONCURRENT_OFF) {
2943                    /* Not in a concurrent cycle so do perform the super class precollect work */
2944                    MM_ParallelGlobalGC::internalPreCollect(env, subSpace, allocDescription, gcCode);
2945            } else {
(gdb) n
2944                    MM_ParallelGlobalGC::internalPreCollect(env, subSpace, allocDescription, gcCode);
(gdb) s
MM_ParallelGlobalGC::internalPreCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, subSpace=0x7ffff0107640, allocDescription=0x7ffff79a0210, gcCode=0) at ParallelGlobalGC.cpp:740
740     {
(gdb) n
741             _cycleState = MM_CycleState();
(gdb) n
742             env->_cycleState = &_cycleState;
(gdb) n
743             env->_cycleState->_gcCode = MM_GCCode(gcCode);
(gdb) n
744             env->_cycleState->_type = _cycleType;
(gdb) n
745             env->_cycleState->_activeSubSpace = subSpace;
(gdb) n
746             env->_cycleState->_collectionStatistics = &_collectionStatistics;
(gdb) n
751             if (!env->_cycleState->_gcCode.isExplicitGC()) {
(gdb) n
752                     if(excessive_gc_normal != _extensions->excessiveGCLevel) {
(gdb) n
758             GC_OMRVMInterface::flushCachesForGC(env);
(gdb) n
759             Assert_MM_true(NULL != env->_cycleState);
(gdb) n
761             _markingScheme->getMarkMap()->setMarkMapValid(false);
(gdb) n
763             if (_extensions->processLargeAllocateStats) {
(gdb) n
768             reportGCCycleStart(env);
(gdb) s
MM_ConcurrentGC::reportGCCycleStart (this=0x7ffff0030f30, env=0x7ffff00426e8) at ConcurrentGC.cpp:178
178     {
(gdb) n
183             _concurrentCycleState = MM_CycleState();
(gdb) n
184             env->_cycleState = &_concurrentCycleState;
(gdb) n
185             env->_cycleState->_type = _cycleType;
(gdb) n
187             MM_ParallelGlobalGC::reportGCCycleStart(env);
(gdb) s
MM_ParallelGlobalGC::reportGCCycleStart (this=0x7ffff0030f30, env=0x7ffff00426e8) at ParallelGlobalGC.cpp:1163
1163    {
(gdb) n
1164            OMRPORT_ACCESS_FROM_ENVIRONMENT(env);
(gdb) n
1167            TRIGGER_J9HOOK_MM_OMR_GC_CYCLE_START(
(gdb) n
1174    }
(gdb) list 1167,1173
1167            TRIGGER_J9HOOK_MM_OMR_GC_CYCLE_START(
1168                    _extensions->omrHookInterface,
1169                    env->getOmrVMThread(),
1170                    omrtime_hires_clock(),
1171                    J9HOOK_MM_OMR_GC_CYCLE_START,
1172                    _extensions->getHeap()->initializeCommonGCData(env, &commonData),
1173                    env->_cycleState->_type);
(gdb) n
MM_ConcurrentGC::reportGCCycleStart (this=0x7ffff0030f30, env=0x7ffff00426e8) at ConcurrentGC.cpp:189
189             env->_cycleState = NULL;
(gdb) n
190     }
(gdb) n
MM_ParallelGlobalGC::internalPreCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, subSpace=0x7ffff0107640, allocDescription=0x7ffff79a0210, gcCode=0) at ParallelGlobalGC.cpp:769
769             reportGCStart(env);
(gdb) s
MM_ParallelGlobalGC::reportGCStart (this=0x7ffff0030f30, env=0x7ffff00426e8) at ParallelGlobalGC.cpp:1199
1199    {
(gdb) n
1200            uintptr_t scavengerCount = 0;
(gdb) n
1201            OMRPORT_ACCESS_FROM_ENVIRONMENT(env);
(gdb) n
1204            scavengerCount = _extensions->scavengerStats._gcCount;
(gdb) n
1207            Trc_MM_GlobalGCStart(env->getLanguageVMThread(), _extensions->globalGCStats.gcCount);
(gdb) n
1208            Trc_OMRMM_GlobalGCStart(env->getOmrVMThread(), _extensions->globalGCStats.gcCount);
(gdb) n
1210            TRIGGER_J9HOOK_MM_OMR_GLOBAL_GC_START(
(gdb) list 1210,1220
1210            TRIGGER_J9HOOK_MM_OMR_GLOBAL_GC_START(
1211                    _extensions->omrHookInterface,
1212                    env->getOmrVMThread(),
1213                    omrtime_hires_clock(),
1214                    J9HOOK_MM_OMR_GLOBAL_GC_START,
1215                    _extensions->globalGCStats.gcCount,
1216                    scavengerCount,
1217                    env->_cycleState->_gcCode.isExplicitGC() ? 1 : 0,
1218                    env->_cycleState->_gcCode.isAggressiveGC() ? 1: 0,
1219                    _bytesRequested);
1220    }
(gdb) p env->_cycleState
$6 = (MM_CycleState *) 0x0
(gdb) n
$sig = ((siginfo_t*) 0x7ffff799efb0); si_addr = (nil)
(gdb) bt
#0  vmSignalHandler (portLibrary=0x7ffff6f9dec0, gpType=4, gpInfo=0x7ffff799ed90, userData=0x64b300) at gphandle.c:797
#1  0x00007ffff67f22aa in structuredSignalHandler (portLibrary=0x7ffff6f9dec0, gpType=4, gpInfo=0x7ffff799ed90, userData=0x64b300) at gphandle.c:723
#2  0x00007ffff60aabe0 in masterSynchSignalHandler (signal=11, sigInfo=0x7ffff799efb0, contextInfo=0x7ffff799ee80) at ../../omr/port/unix/omrsignal.c:838
#3  <signal handler called>
#4  0x00007ffff449fcc4 in MM_GCCode::isExplicitGC (this=0x0) at GCCode.cpp:32
#5  0x00007ffff455e75e in MM_ParallelGlobalGC::reportGCStart (this=0x7ffff0030f30, env=0x7ffff00426e8) at ParallelGlobalGC.cpp:1210
#6  0x00007ffff455ca01 in MM_ParallelGlobalGC::internalPreCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, subSpace=0x7ffff0107640, allocDescription=0x7ffff79a0210, gcCode=0) at ParallelGlobalGC.cpp:769
#7  0x00007ffff454bcbf in MM_ConcurrentGC::internalPreCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, subSpace=0x7ffff0107640, allocDescription=0x7ffff79a0210, gcCode=0) at ConcurrentGC.cpp:2944
#8  0x00007ffff4499ba4 in MM_Collector::preCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, subSpace=0x7ffff0107640, allocDescription=0x7ffff79a0210, gcCode=0) at Collector.cpp:202
#9  0x00007ffff449a059 in MM_Collector::garbageCollect (this=0x7ffff0030f30, env=0x7ffff00426e8, callingSubSpace=0x7ffff0107640, allocateDescription=0x7ffff79a0210, gcCode=0, objectAllocationInterface=0x0, baseSubSpace=0x7ffff0107290, context=0x0) at Collector.cpp:301
#10 0x00007ffff46066c6 in MM_MemorySubSpaceFlat::allocationRequestFailed (this=0x7ffff0107640, env=0x7ffff00426e8, allocateDescription=0x7ffff79a0210, allocationType=MM_MemorySubSpace::ALLOCATION_TYPE_OBJECT, objectAllocationInterface=0x0, baseSubSpace=0x7ffff0107290, previousSubSpace=0x7ffff0107290) at MemorySubSpaceFlat.cpp:143
#11 0x00007ffff4608f92 in MM_MemorySubSpaceGeneric::allocateObject (this=0x7ffff0107290, env=0x7ffff00426e8, allocDescription=0x7ffff79a0210, baseSubSpace=0x0, previousSubSpace=0x0, shouldCollectOnFailure=true) at MemorySubSpaceGeneric.cpp:303
#12 0x00007ffff4617a88 in MM_TLHAllocationInterface::allocateObject (this=0x7ffff0136370, env=0x7ffff00426e8, allocDescription=0x7ffff79a0210, memorySpace=0x7ffff01079b0, shouldCollectOnFailure=true) at TLHAllocationInterface.cpp:194
#13 0x00007ffff4617c6a in MM_TLHAllocationInterface::allocateArrayletSpine (this=0x7ffff0136370, env=0x7ffff00426e8, allocateDescription=0x7ffff79a0210, memorySpace=0x7ffff01079b0, shouldCollectOnFailure=true) at TLHAllocationInterface.cpp:226
#14 0x00007ffff45cf484 in MM_ArrayletAllocationModel::allocateAndInitializeLayout (this=0x7ffff01365a0, memorySpace=0x7ffff01079b0, env=0x7ffff00426e8, allocDesc=0x7ffff79a0210, numberIndexedFields=1310720, shouldCollectOnFailure=true) at ArrayletAllocationModel.cpp:122
#15 0x00007ffff45d0fa0 in MM_ArrayletAllocationModel::allocateIndexableObject (this=0x7ffff01365a0, env=0x7ffff00426e8, allocDescription=0x7ffff79a0210, numberIndexedFields=1310720, shouldCollectOnFailure=true) at ArrayletAllocationModel.cpp:315
#16 0x00007ffff42aad65 in J9AllocateIndexableObject (vmThread=0x64b300, clazz=0x689c00, size=1310720, allocateFlags=1) at mgcalloc.cpp:516
#17 0x00007ffff6768994 in VM_BytecodeInterpreter::anewarray (this=0x7ffff79a0500, _sp=@0x7ffff79a0468, _pc=@0x7ffff79a0470) at BytecodeInterpreter.hpp:7283
#18 0x00007ffff6780678 in VM_BytecodeInterpreter::run (this=0x7ffff79a0500, vmThread=0x64b300) at BytecodeInterpreter.hpp:9846
#19 0x00007ffff674c9c1 in bytecodeLoop (currentThread=0x64b300) at BytecodeInterpreter.cpp:90
#20 0x00007ffff67f795e in cInterpreter () at javamisc.s:222
#21 0x0000000000000000 in ?? ()
(gdb)
```